src/Sentry/Internal/MainExceptionProcessor.cs

```diff
@@ -49,7 +56,30 @@ public void Process(Exception exception, SentryEvent sentryEvent)
 
                 foreach (var key in sentryException.Data.Keys)
                 {
-                    sentryEvent.SetExtra($"Exception[{i}][{key}]", sentryException.Data[key]);
+                    if (key.StartsWith("sentry", StringComparison.OrdinalIgnoreCase))
+                    {
+                        if (key.StartsWith(ExceptionDataTagKey, StringComparison.OrdinalIgnoreCase))
+                        {
+                            var data = sentryException.Data[key] as string ?? "<invalid>";
+                            //Set the key after the ExceptionDataTagKey string.
+                            sentryEvent.SetTag(key.Substring(ExceptionDataTagKey.Length), data);
+                        }
+                        else if (key.StartsWith(ExceptionDataContextKey, StringComparison.OrdinalIgnoreCase))
+                        {
+                            //Set the key after the ExceptionDataTagKey string.
+                            _ = sentryEvent.Contexts.AddOrUpdate(key.Substring(ExceptionDataContextKey.Length),
+                                sentryException.Data[key] ?? "<empty>",
+                                (_, value) => value);
+                        }
+                        else
+                        {
+                            sentryEvent.SetExtra($"Exception[{i}][{key}]", sentryException.Data[key]);
```

Comment:
I believe Exception has a data bag now, this should go there instead. Not on this PR though.

On the mechanism:
https://develop.sentry.dev/sdk/event-payloads/exception/#exception-mechanism

![image](https://user-images.githubusercontent.com/1633368/108556105-7b234e80-72c4-11eb-8276-7702c4eb8bad.png)